apps/gdalalg_vector_sql.cpp

```diff
@@ -0,0 +1,306 @@
+            UserData *pUserData = static_cast<UserData *>(pUserDataIn);
+            pUserData->poSrcDS->ReleaseResultSet(poLayer);
+        };
+
+        const auto DeleteUserData = [](void *pUserDataIn)
+        { delete static_cast<UserData *>(pUserDataIn); };
+
+        auto pUserData = new UserData;
+        pUserData->poSrcDS = m_poSrcDS;
+        pUserData->osDialect = osDialect;
+        pUserData->osSQL = osSQL;
+        pUserData->osLayerName = osLayerName;
+        auto poLayer = std::make_unique<ProxiedSQLLayer>(
+            osLayerName, &m_oPool, OpenLayer, CloseLayer, DeleteUserData,
+            pUserData);
+        m_layers.push_back(std::move(poLayer));
+    }
+
+    int GetLayerCount() override
+    {
+        return static_cast<int>(m_layers.size());
+    }
+
+    OGRLayer *GetLayer(int idx) override
+    {
+        return idx >= 0 && idx < GetLayerCount() ? m_layers[idx].get()
+                                                 : nullptr;
+    }
+};
+}  // namespace
+
+/************************************************************************/
+/*                 GDALVectorSQLAlgorithm::RunStep()                    */
+/************************************************************************/
+
+bool GDALVectorSQLAlgorithm::RunStep(GDALProgressFunc, void *)
+{
+    CPLAssert(m_inputDataset.GetDatasetRef());
+    CPLAssert(m_outputDataset.GetName().empty());
+    CPLAssert(!m_outputDataset.GetDatasetRef());
+
+    if (!m_outputLayer.empty() && m_outputLayer.size() != m_sql.size())
+    {
+        ReportError(CE_Failure, CPLE_AppDefined,
+                    "There should be as many layer names in --output-layer as "
+                    "in --statement");
+        return false;
+    }
+
+    auto poSrcDS = m_inputDataset.GetDatasetRef();
+
+    if (m_sql.size() == 1)
+    {
+        auto outDS = std::make_unique<GDALVectorSQLAlgorithmDataset>(poSrcDS);
+        outDS->SetDescription(poSrcDS->GetDescription());
+
+        OGRLayer *poLayer = poSrcDS->ExecuteSQL(
+            m_sql[0].c_str(), nullptr,
+            m_dialect.empty() ? nullptr : m_dialect.c_str());
+        if (!poLayer)
+            return false;
+
+        if (!m_outputLayer.empty())
+        {
+            const std::string &osLayerName = m_outputLayer[0];
+            poLayer->GetLayerDefn()->SetName(osLayerName.c_str());
+            poLayer->SetDescription(osLayerName.c_str());
+        }
+        outDS->AddLayer(poLayer);
+        m_outputDataset.Set(std::move(outDS));
+    }
+    else
+    {
+        // First pass to check all statements are valid and figure out layer
+        // names
+        std::set<std::string> setOutputLayerNames;
+        std::vector<std::string> aosLayerNames;
+        for (const std::string &sql : m_sql)
+        {
+            auto poLayer = poSrcDS->ExecuteSQL(
+                sql.c_str(), nullptr,
+                m_dialect.empty() ? nullptr : m_dialect.c_str());
+            if (!poLayer)
+                return false;
```

Comment:
error reporting?

Reply:
in most cases a CPLError() should have been emitted by ExecuteSQL(), but thinking deeper, there might be cases where the statement is valid but is not a SELECT-like one and thus return a null layer without error. I've added handling for that case. Thanks.